source/common/sequencer_impl.cc

```diff
@@ -84,36 +82,29 @@ void SequencerImpl::run(bool from_periodic_timer) {
   const auto now = time_source_.monotonicTime();
   const auto running_duration = now - start_time_;
 
-  // If we exceed the benchmark duration.
-  if (cancelled_ || (running_duration > duration_)) {
-    if (targets_completed_ == targets_initiated_) {
-      // All work has completed. Stop this sequencer.
-      stop(false);
-    } else {
-      // After the benchmark duration has passed, we wait for a grace period for outstanding work
-      // to wrap up. If that takes too long we warn about it and quit.
-      if (running_duration - duration_ > grace_timeout_) {
-        stop(true);
-        return;
-      }
-      if (from_periodic_timer) {
-        scheduleRun();
-      }
-    }
-    return;
-  }
-
   // The running_duration we compute will be negative until it is time to start.
   if (running_duration >= 0ns) {
+    last_termination_status_ = last_termination_status_ == TerminationPredicate::Status::PROCEED
+                                   ? termination_predicate_.evaluateChain()
+                                   : last_termination_status_;
+    // If we should stop according to termination conditions.
+    if (last_termination_status_ != TerminationPredicate::Status::PROCEED) {
+      stop(last_termination_status_ == TerminationPredicate::Status::FAIL);
+      return;
+    }
+
     while (rate_limiter_->tryAcquireOne()) {
       // The rate limiter says it's OK to proceed and call the target. Let's see if the target is OK
       // with that as well.
       const bool target_could_start = target_([this, now](bool, bool) {
         const auto dur = time_source_.monotonicTime() - now;
         latency_statistic_->addValue(dur.count());
         targets_completed_++;
-        // Immediately schedule us to check again, as chances are we can get on with the next task.
-        spin_timer_->enableTimer(0ms);
+        if (spin_timer_ != nullptr) {
```

Comment:
comment why might this be null and why is it safe to not schedule?

Reply:
Done